sdk/storage/Azure.Storage.Queues/src/QueueClientSideDecryptor.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Storage.Cryptography;
+using Azure.Storage.Queues.Models;
+using Azure.Storage.Queues.Specialized;
+using Azure.Storage.Queues.Specialized.Models;
+
+namespace Azure.Storage.Queues
+{
+    internal class QueueClientSideDecryptor
+    {
+        private readonly ClientSideDecryptor _decryptor;
+        private readonly IClientSideDecryptionFailureListener _listener;
+
+        public QueueClientSideDecryptor(ClientSideDecryptor decryptor, IClientSideDecryptionFailureListener listener)
+        {
+            _decryptor = decryptor;
+            _listener = listener;
+        }
+
+        public async Task<QueueMessage[]> ClientSideDecryptMessagesInternal(QueueMessage[] messages, bool async, CancellationToken cancellationToken)
+        {
+            var filteredMessages = new List<QueueMessage>();
+            foreach (var message in messages)
+            {
+                try
+                {
+                    message.MessageText = await ClientSideDecryptInternal(message.MessageText, async, cancellationToken).ConfigureAwait(false);
+                    filteredMessages.Add(message);
+                }
+                catch (Exception e) when (_listener != default)
+                {
+                    if (async)
```

Comment:
This pattern doesn't feel right.  What if I implemented a sync listener becuase I only Console.WriteLine but use QueueClient.ReceiveAsync to make an async method call?

Reply:
Will come back to this based on discussion of what form the listener should even take.